src/Raven.Server/Rachis/Follower.cs

```diff
@@ -602,16 +607,36 @@ private void ReadAndCommitSnapshot(TransactionOperationContext context, InstallS
             }
         }
 
-        private unsafe bool InstallSnapshot(TransactionOperationContext context, CancellationToken token)
+        private bool InstallSnapshot(TransactionOperationContext context, CancellationToken token)
         {
             var txw = context.Transaction.InnerTransaction;
-            var reader = _connection.CreateReader();
+
+            var fileName = $"snapshot.{Guid.NewGuid():N}";
+            var filePath = context.Environment.Options.DataPager.Options.TempPath.Combine(fileName);
+            using (var temp = new StreamsTempFile(filePath.FullPath, context.Environment))
+            using(var stream = temp.StartNewStream())
+            {
+                var remoteReader = _connection.CreateReaderToStream(stream);
+                if (ReadSnapshot(remoteReader, context, txw, dryRun:true, token) == false)
+                    return false;
+
+                stream.Seek(0, SeekOrigin.Begin);
+                var fileReader = new StreamReader(stream);
+                ReadSnapshot(fileReader, context, txw, dryRun: false, token);
+            }
+            
+            return true;
+        }
+        
+        private unsafe bool ReadSnapshot(Reader reader, TransactionOperationContext context, Transaction txw, bool dryRun, CancellationToken token)
+        {
+            var type = reader.ReadInt32();
+            if (type == -1)
```

Comment:
Is this correct? This was moved outside of the loop and switch statement doesn't have '-1' case

Reply:
Yes, -1 should be sent just in the beginning and in a case we don't need to send the snapshot
https://github.com/ravendb/ravendb/blob/6787a88d96c10c061e6b3d88525373c39a6e0873/src/Raven.Server/Rachis/FollowerAmbassador.cs#L510

Reply:
I think in this way the meaning is more explicit and in a case, we will get "-1" in the middle we will get more describe error